util/rpcclient/rpcclient.go

```diff
@@ -0,0 +1,208 @@
+	f.String(prefix+".jwtsecret", defaultConfig.JWTSecret, "path to file with jwtsecret for validation - ignored if url is self or self-auth")
+	f.Duration(prefix+".connection-wait", defaultConfig.ConnectionWait, "how long to wait for initial connection")
+	f.Duration(prefix+".timeout", defaultConfig.Timeout, "per-response timeout (0-disabled)")
+	f.Uint(prefix+".arg-log-limit", defaultConfig.ArgLogLimit, "limit size of arguments in log entries")
+	f.Uint(prefix+".retries", defaultConfig.Retries, "number of retries in case of failure(0 mean one attempt)")
+	f.String(prefix+".retry-errors", defaultConfig.RetryErrors, "Errors matching this regular expression are automatically retried")
+}
+
+type RpcClient struct {
+	config    ClientConfigFetcher
+	client    *rpc.Client
+	autoStack *node.Node
+	logId     uint64
+}
+
+func NewRpcClient(config ClientConfigFetcher, stack *node.Node) *RpcClient {
+	return &RpcClient{
+		config:    config,
+		autoStack: stack,
+	}
+}
+
+func (c *RpcClient) Close() {
+	c.client.Close()
+}
+
+func limitString(limit int, str string) string {
+	if limit == 0 || len(str) <= limit {
+		return str
+	}
+	prefix := str[:limit/2-1]
+	postfix := str[len(str)-limit/2+1:]
+	return fmt.Sprintf("%v...%v", prefix, postfix)
+}
+
+func logArgs(limit int, args ...interface{}) string {
+	res := "["
+	for i, arg := range args {
+		res += limitString(limit, fmt.Sprintf("%+v", arg))
+		if i < len(args)-1 {
+			res += ", "
+		}
+	}
+	res += "]"
+	return res
+}
+
+func (c *RpcClient) CallContext(ctx_in context.Context, result interface{}, method string, args ...interface{}) error {
+	if c.client == nil {
+		return errors.New("not connected")
+	}
+	logId := atomic.AddUint64(&c.logId, 1)
+	log.Trace("sending RPC request", "method", method, "logId", logId, "args", logArgs(int(c.config().ArgLogLimit), args...))
+	var err error
+	for i := 0; i < int(c.config().Retries)+1; i++ {
+		if ctx_in.Err() != nil {
+			return ctx_in.Err()
+		}
+		var ctx context.Context
+		var cancelCtx context.CancelFunc
+		timeout := c.config().Timeout
+		if timeout > 0 {
+			ctx, cancelCtx = context.WithTimeout(ctx_in, timeout)
+		} else {
+			ctx, cancelCtx = context.WithCancel(ctx_in)
+		}
+		err = c.client.CallContext(ctx, result, method, args...)
+		cancelCtx()
+		logger := log.Trace
+		limit := int(c.config().ArgLogLimit)
+		if err != nil && err.Error() != "already known" {
+			logger = log.Info
+			limit = 0
+		}
+		logger("rpc response", "method", method, "logId", logId, "result", limitString(limit, fmt.Sprintf("%+v", result)), "attempt", i, "args", logArgs(limit, args...))
+		if err == nil {
+			return nil
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
+			continue
+		}
+		retryErrors := c.config().RetryErrors
+		if retryErrors != "" {
+			match, regexErr := regexp.MatchString(retryErrors, err.Error())
+			if regexErr != nil {
+				log.Warn("rpcclient: bad value for retry-error. Not retrying.", "err", err, "value", retryErrors)
```

Comment:
I don't think we _need_ to do this for this PR, but this would be a good use-case for caching with a private field of the config. We could cache the compiled retryErrors regex in a private field of the RPC client config which gets populated on config validation with the Validate method.